controller/api/destination/server.go

```diff
@@ -344,58 +324,132 @@ func (s *server) GetProfile(dest *pb.GetDestination, stream pb.Destination_GetPr
 // If the pod argument is provided, the profile sent to the client will
 // include an endpoint. Otherwise, the default profile is sent.
 func (s *server) sendEndpointProfile(stream pb.Destination_GetProfileServer, pod *corev1.Pod, port uint32) error {
-	log := s.log
-	var endpoint *pb.WeightedAddr
-	opaquePorts := make(map[uint32]struct{})
-	var err error
-	if pod != nil {
-		podSet := podToAddressSet(s.k8sAPI, pod).WithPort(port)
-		podID := watcher.PodID{
-			Namespace: pod.Namespace,
-			Name:      pod.Name,
-		}
-		var ok bool
-		opaquePorts, ok, err = getPodOpaquePortsAnnotations(pod)
-		if err != nil {
-			log.Errorf("failed to get opaque ports annotation for pod: %s", err)
-		}
+	if pod == nil {
+		translator := newProfileTranslator(stream, s.log, "", port, nil)
+		translator.Update(nil)
+		return nil
+	}
 
-		// If the opaque ports annotation was not set, then set the
-		// endpoint's opaque ports to the default value.
-		if !ok {
-			opaquePorts = s.defaultOpaquePorts
-		}
+	opaquePorts, ok, err := getPodOpaquePortsAnnotations(pod)
+	if err != nil {
+		s.log.Errorf("failed to get opaque ports annotation for pod: %s", err)
+	}
 
-		skippedInboundPorts, err := getPodSkippedInboundPortsAnnotations(pod)
-		if err != nil {
-			log.Errorf("failed to get ignored inbound ports annotation for pod: %s", err)
+	// If the opaque ports annotation was not set, then use the destination
+	// server's default value.
+	if !ok {
+		opaquePorts = s.defaultOpaquePorts
+	}
+
+	skippedInboundPorts, err := getPodSkippedInboundPortsAnnotations(pod)
+	if err != nil {
+		s.log.Errorf("failed to get ignored inbound ports annotation for pod: %s", err)
+	}
+	podSet := podToAddressSet(s.k8sAPI, pod).WithPort(port)
+	podID := watcher.PodID{
+		Namespace: pod.Namespace,
+		Name:      pod.Name,
+	}
+	endpoint, err := toWeightedAddr(podSet.Addresses[podID], opaquePorts, skippedInboundPorts, s.enableH2Upgrade, s.identityTrustDomain, s.controllerNS, s.log)
+	if err != nil {
+		return err
+	}
+
+	// `Get` doesn't include the namespace in the per-endpoint metadata, so it
+	// needs to be special-cased.
+	endpoint.MetricLabels["namespace"] = pod.Namespace
+
+	translator := newProfileTranslator(stream, s.log, "", port, endpoint)
+
+	// Create a client which watches for changes to the inbound policy for
+	// portSpec.
+	portSpec := &policyPb.PortSpec{
+		Workload: fmt.Sprintf("%s:%s", pod.Namespace, pod.Name),
+		Port:     port,
+	}
+	client, err := s.policyClient.WatchPort(stream.Context(), portSpec)
```

Comment:
It's worth noting that this will create one WatchPort policy request for each GetProfile request.  This is probably okay for now, though for large clusters this could potentially create a very large number of redundant watches to the policy controller.